### PR TITLE
GLT-2213: allow users to delete Dilutions

### PR DESCRIPTION
**Pull request by djcooke**
More avoiding existing SecurityProfile issues - very few SecurityProfiles actually have an owner set, so I removed that requirement for inheriting. Most securable entities have a `creator` field that has been used instead.

I've created a separate ticket (GLT-2220) for storing `userId` instead of `fullName` as the dilution creator.

### CODE REVIEW

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/security/SecurableByProfile.java [hunk 1]
```diff
@@ -56,7 +56,7 @@
    *           when the parent SecurityProfile could not be inherited
    */
   public default void inheritPermissions(SecurableByProfile parent) throws SecurityException {
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/security/SecurableByProfile.java]: MINOR Remove the declaration of thrown exception 'java.lang.SecurityException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

**Comment by sqwci** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [LibraryDilution.java#L343](https://github.com/TGAC/miso-lims/blob/84940feabea168ee5a8bafd0b5fc371661915b8c/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/LibraryDilution.java#L343): Update this method so that its implementation is not identical to "getName" on line 172. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4144)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by djcooke** [comment]
@callunity: fixed. The DB stored procedure doesn't affect this, but `DeleterService#bulkDelete` was calling before/afterDelete methods and its own `delete` method, which also calls before/afterDelete

**Comment by sqwci** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [LibraryDilution.java#L343](https://github.com/TGAC/miso-lims/blob/a9858f4dcfcc7636b75a6dc4193a61a01582008f/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/LibraryDilution.java#L343): Update this method so that its implementation is not identical to "getName" on line 172. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4144)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultLibraryDilutionService.java [hunk 2]
```diff
@@ -256,4 +277,39 @@ public AuthorizationManager getAuthorizationManager() {
...
+  @Override
+  public void authorizeDeletion(LibraryDilution object) throws IOException {
+    User creator = securityStore.getUserByFullName(object.getDilutionCreator());
+    authorizationManager.throwIfNonAdminOrMatchingOwner(creator);
+  }
+
+  @Override
+  public ValidationResult validateDeletion(LibraryDilution object) {
+    ValidationResult result = new ValidationResult();
+
```
Comment (apmasell) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultLibraryDilutionService.java]:
```
    if (object.getBox() != null) {
      result.addError(new ValidationError(
          String.format("%s is in box %s (%s)", object.getName(), object.getBox().getName(), object.getBox().getAlias())));
    }
```

**Comment by apmasell** [comment]
This needs to be changed to match #1366 to handle the deletion of boxed items.

**Comment by sqwci** [comment]
SonarQube analysis reported 6 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [DefaultSampleService.java#L226](https://github.com/TGAC/miso-lims/blob/06d0f7b1f99f21f8acc724b7de515da100928be0/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L226): Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [LibraryDilution.java#L343](https://github.com/TGAC/miso-lims/blob/06d0f7b1f99f21f8acc724b7de515da100928be0/core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/impl/LibraryDilution.java#L343): Update this method so that its implementation is not identical to "getName" on line 172. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS4144)
1. ![MAJOR][MAJOR] [DefaultSampleService.java#L90](https://github.com/TGAC/miso-lims/blob/06d0f7b1f99f21f8acc724b7de515da100928be0/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L90): A "NullPointerException" could be thrown; "parent" is nullable here. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2259)
1. ![MINOR][MINOR] [DefaultSampleService.java#L416](https://github.com/TGAC/miso-lims/blob/06d0f7b1f99f21f8acc724b7de515da100928be0/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L416): Remove the declaration of thrown exception 'org.hibernate.exception.ConstraintViolationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [DefaultSampleService.java#L506](https://github.com/TGAC/miso-lims/blob/06d0f7b1f99f21f8acc724b7de515da100928be0/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultSampleService.java#L506): Remove the declaration of thrown exception 'java.sql.SQLException', as it cannot be thrown from method's body. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by djcooke** [comment]
@apmasell: updated `LibraryDilutionService` to implement `BoxableDeleterService`

**Pull request merged**